pkg/controllers/machine/link/controller.go

```diff
@@ -0,0 +1,129 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package link
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/patrickmn/go-cache"
+	"github.com/samber/lo"
+	"go.uber.org/multierr"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/aws/karpenter/pkg/cloudprovider"
+
+	"github.com/aws/karpenter-core/pkg/apis/v1alpha5"
+	corecloudprovider "github.com/aws/karpenter-core/pkg/cloudprovider"
+	"github.com/aws/karpenter-core/pkg/operator/controller"
+	machineutil "github.com/aws/karpenter-core/pkg/utils/machine"
+)
+
+type Controller struct {
+	kubeClient    client.Client
+	cloudProvider *cloudprovider.CloudProvider
+	cache         *cache.Cache // exists due to eventual consistency on the controller-runtime cache
+}
+
+func NewController(kubeClient client.Client, cloudProvider *cloudprovider.CloudProvider) controller.Controller {
+	return &Controller{
+		kubeClient:    kubeClient,
+		cloudProvider: cloudProvider,
+		cache:         cache.New(time.Minute, time.Second*10),
+	}
+}
+
+func (c *Controller) Name() string {
+	return "machine.link"
+}
+
+func (c *Controller) Reconcile(ctx context.Context, _ reconcile.Request) (reconcile.Result, error) {
+	machineList := &v1alpha5.MachineList{}
+	if err := c.kubeClient.List(ctx, machineList); err != nil {
+		return reconcile.Result{}, err
+	}
+	retrieved, err := c.cloudProvider.List(ctx)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("listing cloudprovider machines, %w", err)
+	}
+	// Filter out any machines that shouldn't be linked
+	retrieved = lo.Filter(retrieved, func(m *v1alpha5.Machine, _ int) bool {
+		return c.shouldLink(m, machineList.Items)
+	})
+	errs := make([]error, len(retrieved))
+	workqueue.ParallelizeUntil(ctx, 20, len(retrieved), func(i int) {
+		errs[i] = c.link(ctx, retrieved[i])
+	})
+	// Effectively, don't requeue this again once it succeeds
+	return reconcile.Result{RequeueAfter: math.MaxInt64}, multierr.Combine(errs...)
+}
+
+func (c *Controller) link(ctx context.Context, retrieved *v1alpha5.Machine) error {
+	provisionerName, ok := retrieved.Labels[v1alpha5.ProvisionerNameLabelKey]
+	if !ok {
+		return corecloudprovider.IgnoreMachineNotFoundError(c.cloudProvider.Delete(ctx, retrieved))
+	}
+	provisioner := &v1alpha5.Provisioner{}
+	if err := c.kubeClient.Get(ctx, types.NamespacedName{Name: provisionerName}, provisioner); err != nil {
+		if errors.IsNotFound(err) {
+			return corecloudprovider.IgnoreMachineNotFoundError(c.cloudProvider.Delete(ctx, retrieved))
+		}
+		return err
+	}
+	machine := machineutil.New(&v1.Node{}, provisioner)
```

Comment:
why does this take a pointer to a node?

Reply:
This was originally there so that the machineutil.New() func could use the NewFromNode for status details and then layer that with provisioner spec details. There may be room to clean this up later but I think it's reasonable to keep it for now